Adding ExVCR addon

hanam1ni opened the pull request:
## What happened

As mentioned in [Compass](https://nimblehq.co/compass/development/code-conventions/ex-unit/#stubbing-network-requests), we going to use ExVCR for stubbing network request in tests. In this PR, adding ExVCR addon:

- Adding ExVCR dependency. 
- Update ExVCR configuration. 
- Creating a directory for cassettes.
 
## Insight

Adding 2 helper functions in `Nimble.Phx.Gen.Template.Generator` module

- `append_content/2` for appending content to the given file.
- `make_directory/2` for creating directory and providing an option to create a `.keep` file.
 
## Proof Of Work

After applied templated to the project:

- ExVCR is added in `deps` of `mix.exs`
![image](https://user-images.githubusercontent.com/14077479/113559288-47f60e00-962b-11eb-88f8-a3957309a709.png)

- Configuration is updated in `config/test.exs`
![image](https://user-images.githubusercontent.com/14077479/113559349-6825cd00-962b-11eb-923f-8a97847d3e1a.png)

- `DataCase` and `ConnCase` are updated
![image](https://user-images.githubusercontent.com/14077479/113559199-2a28a900-962b-11eb-8ed7-f952a4e98b1c.png)

- Create a dummy module and add a test. After running the test, a cassette is created.
![image](https://user-images.githubusercontent.com/14077479/113559466-9a372f00-962b-11eb-9ff5-67effd62f9a9.png)

byhbt commented on lib/nimble.phx.gen.template/addons/ex_vcr.ex:
```diff
@@ -0,0 +1,73 @@
+  @cassette_directory "test/support/fixtures/vcr_cassettes"
+
+  @impl true
+  def do_apply(%Project{} = project, _opts) do
```
I have a question here about this variable `_opts`, how to deal with the unused variable? if we don't use it should we rename it to `_` ? 🤔

hanam1ni replied:
For unused variables, we can name with underscore or prefix underscore with the name. In this project, we set Credo to always name unused variables with an underscore for clarification.

byhbt replied:
thanks @hanam1ni !

byhbt commented on lib/nimble.phx.gen.template/generator.ex:
```diff
@@ -86,4 +111,16 @@ defmodule Nimble.Phx.Gen.Template.Generator do
+        :ok
+
+      {:error, reason} ->
+        Mix.raise(~s[Failed to create keep file at #{path} reason: #{Atom.to_string(reason)}])
```
```suggestion
        Mix.raise(~s[Failed to create .keep file at #{path}. Reason: #{Atom.to_string(reason)}])
```

hanam1ni replied:
Updated the error message and use comma as a separator.

byhbt commented on test/nimble.phx.gen.template/addons/ex_vcr_test.exs:
```diff
@@ -0,0 +1,62 @@
+  describe "#apply/2" do
+    @describetag mock_latest_package_versions: [{:exvcr, "0.12.2"}]
+
+    test "injects exvcr to mix dependency", %{
```
```suggestion
    test "injects ExVCR to mix dependency", %{
```

Update the case of `excvr` for the test description to make it easier to read.

andyduong1920 commented:
Should we made this as an addon @hanam1ni because not all project need to use ExVCR 🤔

hanam1ni commented:
> Should we make this as an optianl addon @hanam1ni because not all project need to use ExVCR 🤔

@andyduong1920 Updated to make it optional
![image](https://user-images.githubusercontent.com/14077479/113656784-c227a080-96c6-11eb-8363-7aec08a089d4.png)

andyduong1920 commented:
@hanam1ni The PR looks great since we made the ExVCR addon optional now so I think it better to test that addon on the integration test too, what do you think? we can do that by adding the `Y` answer in the Makefile https://github.com/nimblehq/elixir-templates/blob/develop/Makefile#L13

```
# Y - in response to Would you like to add the Oban addon?
# Y - in response to Would you like to add the ExVCR addon?
common_addon_prompts = Y\nY\nY\nY\nY\n
```

andyduong1920 commented:
Since we have merged the Mix variant, what do you think if we could add ExVCR addon into the Mix template too @hanam1ni 🙏

hanam1ni commented:
@andyduong1920 I updated the `Makefile` and refactored the modules, but I don't think we should add `ExVCR` in the mix template. 

To make the ExVCR work, we need to update the `config` and `cases` modules but the Mix project does not generate those by default. So I think we should not add `ExVCR` in the Mix template to make the project minimal.

andyduong1920 commented:
> To make the ExVCR work, we need to update the `config` and `cases` modules but the Mix project does not generate those by default. So I think we should not add `ExVCR` in the Mix template to make the project minimal.

Yeah, agreed @hanam1ni
